File: server-go/internal/themes/themes_test.go
package themes

import (
	"testing"
	"time"
)

func TestRegistryStructure(t *testing.T) {
	if len(Registry) < 13 {
		t.Fatalf("expected at least 13 themes, got %d", len(Registry))
	}
	seen := map[string]bool{}
	for _, th := range Registry {
		if th.Slug == "" {
			t.Errorf("empty slug in theme %q", th.Name)
		}
		if seen[th.Slug] {
			t.Errorf("duplicate slug %q", th.Slug)
		}
		seen[th.Slug] = true
		if len(th.Windows) == 0 {
			t.Errorf("theme %q has no window", th.Slug)
		}
		if th.Decoration.LogoBadge == "" {
			t.Errorf("theme %q missing logoBadge", th.Slug)
		}
	}
}

func TestActiveAt(t *testing.T) {
	cases := []struct {
		label     string
		date      time.Time
		wantSlug  string
		wantNotIn string
	}{
		{"圣诞节", time.Date(2026, 12, 25, 10, 0, 0, 0, time.UTC), "christmas", "halloween"},
		{"跨年 1.1", time.Date(2027, 1, 1, 3, 0, 0, 0, time.UTC), "christmas", "halloween"},
		{"万圣节", time.Date(2026, 10, 31, 10, 0, 0, 0, time.UTC), "halloween", "christmas"},
		// 2026 端午 = 公历 6-19;2026 中秋 = 公历 9-25
		{"端午节", time.Date(2026, 6, 19, 10, 0, 0, 0, time.UTC), "dragon-boat", "halloween"},
		{"中秋节", time.Date(2026, 9, 25, 10, 0, 0, 0, time.UTC), "mid-autumn", "halloween"},
		{"国庆节", time.Date(2026, 10, 3, 10, 0, 0, 0, time.UTC), "national-day", "christmas"},
		{"植树节", time.Date(2026, 3, 12, 10, 0, 0, 0, time.UTC), "arbor-day", "christmas"},
	}
	for _, c := range cases {
		t.Run(c.label, func(t *testing.T) {
			got := ActiveAt(c.date)
			found := false
			notFound := true
			for _, th := range got {
				if th.Slug == c.wantSlug {
					found = true
				}
				if th.Slug == c.wantNotIn {
					notFound = false
				}
			}
			if !found {
				t.Errorf("expected %q active on %v, got %v", c.wantSlug, c.date, slugs(got))
			}
			if !notFound {
				t.Errorf("%q should NOT be active on %v", c.wantNotIn, c.date)
			}
		})
	}
}

func TestInactive(t *testing.T) {
	// 5 月 7 日既非社区生日(示例是 5.1-5.3)也非任何其他节日
	d := time.Date(2026, 5, 7, 10, 0, 0, 0, time.UTC)
	got := ActiveAt(d)
	for _, th := range got {
		if th.Slug == "halloween" || th.Slug == "christmas" {
			t.Errorf("%q should not be active on 5/7, got %v", th.Slug, d)
		}
	}
}

func TestInWindowCrossYear(t *testing.T) {
	w := Window{StartMonth: 12, StartDay: 20, EndMonth: 1, EndDay: 2}
	cases := []struct {
		m, d int
		want bool
	}{
		{12, 19, false},
		{12, 20, true},
		{12, 31, true},
		{1, 1, true},
		{1, 2, true},
		{1, 3, false},
	}
	for _, c := range cases {
		if got := inWindow(c.m, c.d, w); got != c.want {
			t.Errorf("inWindow(%d-%d) = %v, want %v", c.m, c.d, got, c.want)
		}
	}
}

func slugs(ts []*Theme) []string {
	out := make([]string, 0, len(ts))
	for _, t := range ts {
		out = append(out, t.Slug)
	}
	return out
}

File: tests/themes.test.mjs
// 使用 node:test(Node 20 内置)跑的轻量单元测试。
//
// 这些测试读 src/lib/themes.ts 的语义(等效 TS/JS),
// 而不是直接 import TS,因此用纯 JS 复制最小等效实现做黑盒验证。
// 目的:验证 TS 端窗口判定与 Go 端一致。

import { test } from 'node:test';
import assert from 'node:assert/strict';

// ---- 最小等效实现(应与 src/lib/themes.ts 一致) ----
function inWindow(m, d, w) {
  const start = w.startMonth * 100 + w.startDay;
  const end = w.endMonth * 100 + w.endDay;
  const cur = m * 100 + d;
  if (start <= end) return cur >= start && cur <= end;
  return cur >= start || cur <= end;
}
function active(w, date) {
  const m = date.getMonth() + 1;
  const d = date.getDate();
  return inWindow(m, d, w);
}

// ---- 测试用例 ----

test('同年窗口 — 春节 1.20 - 2.15', () => {
  const w = { startMonth: 1, startDay: 20, endMonth: 2, endDay: 15 };
  assert.equal(active(w, new Date(2026, 0, 19)), false);
  assert.equal(active(w, new Date(2026, 0, 20)), true);
  assert.equal(active(w, new Date(2026, 1, 15)), true);
  assert.equal(active(w, new Date(2026, 1, 16)), false);
});

test('跨年窗口 — 圣诞节 12.20 - 1.2', () => {
  const w = { startMonth: 12, startDay: 20, endMonth: 1, endDay: 2 };
  assert.equal(active(w, new Date(2026, 11, 19)), false);
  assert.equal(active(w, new Date(2026, 11, 20)), true);
  assert.equal(active(w, new Date(2026, 11, 31)), true);
  assert.equal(active(w, new Date(2027, 0, 1)), true);
  assert.equal(active(w, new Date(2027, 0, 2)), true);
  assert.equal(active(w, new Date(2027, 0, 3)), false);
});

test('单日窗口 — 儿童节 6.1', () => {
  const w = { startMonth: 6, startDay: 1, endMonth: 6, endDay: 1 };
  assert.equal(active(w, new Date(2026, 5, 1)), true);
  assert.equal(active(w, new Date(2026, 4, 31)), false);
  assert.equal(active(w, new Date(2026, 5, 2)), false);
});
